persistence/src/main/java/org/dswarm/persistence/service/internal/graph/InternalGDMGraphService.java

```diff
@@ -552,6 +553,120 @@ public void deleteObject(final String dataModelUuid) throws DMPPersistenceExcept
+
+		final org.dswarm.graph.json.Model realModel = gdmModel.getModel();
+
+		if (realModel == null) {
+
+			throw new DMPPersistenceException("real model that should be added to DB shouldn't be null");
+		}
+
+		final String dataModelURI = GDMUtil.getDataModelGraphURI(dataModelUuid);
+
+		final DataModel finalDataModel = determineSchema(dataModelUuid, gdmModel, realModel);
+
+		final Optional<ContentSchema> optionalContentSchema = Optional.fromNullable(finalDataModel.getSchema().getContentSchema());
+
+		final Optional<Boolean> optionalDeprecateMissingRecords = determineMissingRecordsFlag(updateFormat);
+		final Optional<String> optionalRecordClassUri = Optional.fromNullable(finalDataModel.getSchema().getRecordClass().getUri());
+
+		writeGDMToDB(realModel, dataModelURI, optionalContentSchema, optionalDeprecateMissingRecords, optionalRecordClassUri, enableVersioning);
+	}
+
+	private DataModel determineSchema(final String dataModelUuid, final GDMModel gdmModel, final org.dswarm.graph.json.Model realModel) throws DMPPersistenceException {
+
+		LOG.debug("determine schema for data model '{}'", dataModelUuid);
+
+		final DataModel dataModel = addRecordClass(dataModelUuid, gdmModel.getRecordClassURI());
+
+		final DataModel finalDataModel;
+
+		if (dataModel != null) {
+
+			finalDataModel = dataModel;
+		} else {
+
+			finalDataModel = getDataModel(dataModelUuid);
+		}
+
+		if (finalDataModel == null) {
+
+			throw new DMPPersistenceException("Could not get the actual data model to use");
+		}
+
+		determineRecordResources(gdmModel, realModel, finalDataModel);
+
+		determineAttributePaths(finalDataModel, gdmModel);
+
+		LOG.debug("determined schema for data model '{}'", dataModelUuid);
+
+		return finalDataModel;
+	}
+
+	private void determineRecordResources(final GDMModel gdmModel, final org.dswarm.graph.json.Model realModel, final DataModel finalDataModel) {
+
+		LOG.debug("determine record resources for data model '{}'", finalDataModel.getUuid());
+
+		if (finalDataModel.getSchema() != null) {
+
+			if (finalDataModel.getSchema().getRecordClass() != null) {
+
+				//final String recordClassURI = finalDataModel.getSchema().getRecordClass().getUri();
+
+				// TODO: this operation is expensive, so we should only apply it, if really necessary
+				//final Set<Resource> recordResources = GDMUtil.getRecordResources(recordClassURI, realModel);
+
+//				if (recordResources != null && !recordResources.isEmpty()) {
+//
+//					final LinkedHashSet<String> recordURIs = Sets.newLinkedHashSet();
+//
+//					recordURIs.addAll(recordResources.stream().map(Resource::getUri).collect(Collectors.toList()));
+//
+//					gdmModel.setRecordURIs(recordURIs);
+//				}
+
+				gdmModel.setRecordURIs(realModel.getResourceURIs());
+			}
+		}
+
+		LOG.debug("determined record resources for data model '{}'", finalDataModel.getUuid());
+	}
+
+	private Optional<Boolean> determineMissingRecordsFlag(final UpdateFormat updateFormat) throws DMPPersistenceException {
```

Comment:
Optional<Boolean> seems like overkill, couldn't this just be boolean? You never return Optional[None], anyway.